### ISSUE TITLE
Implement DPCF-like contact hardening to reduce acne and improve shadow rendering quality

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#4661]
### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

Shadow map rendering was already significantly improved in Godot 4.0. Most shadow acne situations can now be avoided without introducing much visible peter-panning. However, we can still go further to improve quality without increasing the performance demands too much.

Additionally, Godot 4 offers percentage-closer soft shadows (PCSS) when full contact hardening of shadows is desired. This can look great, but it's also expensive. Its use should be limited to a few lights per scene at most, and it should be disabled when targeting lower-end GPUs.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Dilated percentage closer filtering (DPCF) is a cheaper implementation of PCSS. It's mainly intended to reduce shadow acne in situations where it can still be noticed, even after the changes made to shadow rendering in Godot 4. DPCF is not intended to provide full contact hardening like PCSS, although it does provide *some* form of contact hardening which may be good enough in some cases.

The principle goes as follows:

- If the depth difference between the caster and the receiving surface is low, decrease the blur radius compared to what it would be otherwise.
- If the depth difference between the caster and receiving surface is high, keep the blur as-is.
- Shadow acne is naturally reduced, as the caster and receiving surface will be very close to each other in this situation. Lower shadow blur will make acne less visible, so it works very well in this situation.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This can likely be implemented in both Vulkan Clustered and Mobile backends, and maybe even the OpenGL backend.

Depending on the performance cost of DPCF, we may want to add a project setting to toggle this (probably enabled by default on desktop, and disabled by default on mobile).

Slides 61 to 74 of https://research.activision.com/publications/2021/10/shadows-of-cold-war--a-scalable-approach-to-shadowing mention a DPCF implementation. Including screenshots of the most important two slides for reference, in case that link goes down:

![image](https://user-images.githubusercontent.com/180032/173243506-85ca3d0f-9389-415d-88fc-ebc01ae7972e.png)

![image](https://user-images.githubusercontent.com/180032/173243536-bc173a6e-2125-4eac-84bb-bb431c31aaca.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

Shadow map rendering is core engine functionality. It can't be overridden by an add-on.

**Comment by VantaGhost** [comment]
I've been looking for some options to make shadows more grounded, such as contact shadows. But this seems like a relatively straightforward enhancement to the current shadowing system that would greatly improve fidelity.
![image](https://github.com/godotengine/godot-proposals/assets/47618902/4e717e84-9788-498c-91ca-40c38e58f441)
I mocked up this image by combining 2 different blur levels. Is this a good approximation of what the effect would look like?

**Comment by Calinou** [comment]
@VantaGhost Yes, but in the case of Blur = 1, you can probably get better bias by manually decreasing bias. The automatic bias adjustment depending on shadow blur can sometimes increaes bias too much (it's conservative and aims to avoid acne as much as possible).
